introduce a function concept

mlubin opened the pull request:
@blegat's comment [here](https://github.com/JuliaOpt/MathOptInterface.jl/issues/9#issuecomment-312516818) exposes the fact that we currently have 6 types of functions with different representations, between the combinations of variablewise, affine, quadratic, and scalar-valued, and vector-valued. Instead of enumerating all of these combinations in ``addconstraint!``, I think it makes sense to encode these with function objects. 

With this, the signatures of ``addconstraint!`` and ``setobjective!`` become:
```julia
addconstraint!(m::AbstractSolverInstance, func::F, set::S)::ConstraintReference{F,S}
setobjective!(m::AbstractSolverInstance, func::F, N::Int=1)
```
where `func` is a simple `struct` with the corresponding data.

``func`` could be, e.g.:
```julia
"""
    ScalarVariablewiseFunction(variable)

The function that extracts the scalar variable referenced by `variable`, a `VariableReference`.
This function would naturally be used for single variable bounds or integrality constraints.
"""
struct ScalarVariablewiseFunction <: AbstractFunction
    variable::VariableReference
end

"""
    VectorQuadraticFunction{T}(affine_variables, affine_coefficients, quadratic_rowvariables, quadratic_colvariables, quadratic_coefficients, constant)

The vector-valued quadratic function with i`th` component ("output index") defined as ``\\frac{1}{2}x^TQ_ix + a_i^T x + b_i``, where:
* ``a_i`` is a sparse vector specified in tuple form by the subset of `affine_variables, affine_coefficients` for the indices `k` where `affine_outputindex[k] == i`.
* ``b_i`` is a scalar specified by `constant[i]`
* ``Q_i`` is a symmetric matrix is specified in triplet form by the subset of `quadratic_rowvariables, quadratic_colvariables, quadratic_coefficients` for the indices `k` where `quadratic_outputindex[k] == i`

Duplicate indices in ``a_i`` or ``Q_i`` are accepted, and the corresponding coefficients are summed together. "Mirrored" indices `(q,r)` and `(r,q)` (where `r` and `q` are `VariableReferences`) are considered duplicates; only one need be specified.
"""
struct VectorQuadraticFunction{T} <: AbstractFunction
    affine_outputindex::Vector{Int}
    affine_variables::Vector{VariableReference}
    affine_coefficients::Vector{T}
    quadratic_outputindex::Vector{Int}
    quadratic_rowvariables::Vector{VariableReference}
    quadratic_colvariables::Vector{VariableReference}
    quadratic_coefficients::Vector{T}
    constant::Vector{T}
end
```

It's still a TODO to rethink the modification and query routines under this function concept, but I believe we'll be able to unify querying and modification across constraints and objectives.

I think this change is already justified as a simplification of the current interface as it is, but in addition to that it also makes everything more extensible for other kinds of functions as requested by @chriscoey.

@joehuchette @ccoffrin @odow @joaquimg

joaquimg commented:
I like it!
People usually add constraints after the complete constraint is well defined. However, it is very common to query and modify very specific parts of a constraint.

Having AbstractFunctions would also be nice for the NLP interface right?

mlubin commented:
> Having AbstractFunctions would also be nice for the NLP interface right?

It's not obvious how to best deal with functions defined by callbacks, but yes, in principle we can use this same interface for NLP.

mlubin commented:
Updated with a new concept of ``AbstractFunctionModification``, so that you can do things like:
```julia
modifyobjective!(m, ScalarConstantChange(10.0))
modifyconstraint!(m, c, ScalarConstantChange(10.0))
```
We no longer need to enumerate all of the possible ways to modify an objective or constraint in ``modifyobjective!`` and ``modifyconstraint!``. Instead, we maintain a list of function modification types. This seems much simpler to me.

I also removed the concept of multiple objectives because the alternative would be passing through objective references everywhere. Nobody has actually requested multiple objective support, nor do I have any particular motivation to support it, so let's simplify things for now.

ccoffrin commented:
I have applications that are multi-objective, however, I usually address them at the JuMP level, not MOI.  So I am in favor of removing this, unless, some solvers have specialized native support detailing with multi-objective.  In that case, I would put it on the wish list for v0.2 of MOI.

mlubin commented:
It's fair to leave multiobjective on the wish list for MOI (Gurobi supports two objectives), but it feels like one step too far right now.

mlubin commented:
This PR is mostly coherent now, so I'll merge soon barring major objections.
